From 05391989747e3c6044329002f0786c37f34a2f6c Mon Sep 17 00:00:00 2001
From: Alan Coopersmith <alan.coopersmith@oracle.com>
Date: Tue, 29 Dec 2015 15:51:45 -0800
Subject: [PATCH 07/19] solaris-notty

Workaround the Solaris PAM & auditing subsystems requirements that
every session cannot be attributed to a single device file, as if
we all still logged in on serial terminals.

Original date:2009-10-16 owner:yippi type:branding
---

--- a/daemon/gdm-manager.c	2024-01-17 22:25:22.072124154 -0800
+++ b/daemon/gdm-manager.c	2024-01-17 22:43:32.618519859 -0800
@@ -29,6 +29,11 @@
 #include <sys/stat.h>
 #include <sys/types.h>
 
+#ifdef __sun
+#include <utime.h>
+#include <sys/param.h>
+#endif
+
 #include <glib.h>
 #include <glib/gi18n.h>
 #include <glib/gstdio.h>
@@ -1155,6 +1160,47 @@
         return recorded;
 }
 
+static void                                  
+create_device (const char *dev)              
+{                                            
+#ifdef __sun                                 
+        gchar buf[MAXPATHLEN + 1];
+        struct stat st;
+        
+        if (dev == NULL || dev[0] == '\0')
+                return;
+        
+        if (strcmp (dev, "/dev/dtlocal")  != 0 &&
+            strcmp (dev, "/dev/dtremote") != 0 )
+                return;
+
+        memset (buf, 0, sizeof (gchar) * (MAXPATHLEN + 1));
+
+        if (stat (dev, &st) != 0) {
+                g_debug ("Creating pseudo-device %s", dev);
+                symlink ("/dev/null", dev);
+        } else if (readlink (dev, buf, MAXPATHLEN) > 0) {
+                if (strcmp (buf, "/dev/null") == 0) {
+                        /* Touch symlink */
+                        struct utimbuf  timebuf;
+        
+                        timebuf.modtime = time ((time_t *) 0);
+                        timebuf.actime  = timebuf.modtime;
+
+                        if ((utime (dev, &timebuf)) != 0)
+                                g_debug ("Problem updating access time of pseudo-device %s", dev);
+                        else
+                                g_debug ("Touching pseudo-device %s", dev);
+                } else {
+                        g_debug ("Device %s points to %s", dev, buf);
+                }
+        } else {
+                g_debug ("Device %s is not a symlink", dev);
+        }
+#endif
+}
+
+
 static GdmSession *
 find_user_session_for_display (GdmManager *self,
                                GdmDisplay *display)
@@ -2810,6 +2856,9 @@
                       "supported-session-types", &supported_session_types,
                       NULL);
         display_device = get_display_device (manager, display);
+        if (!display_device && !display_is_local)
+                display_device = g_strdup ("/dev/dtremote");
+        create_device (display_device);
 
         session = gdm_session_new (GDM_SESSION_VERIFICATION_MODE_LOGIN,
                                    allowed_user,
